pyteal/compiler/optimizer/optimizer_test.py

```diff
@@ -0,0 +1,363 @@
+from pickletools import optimize
+from .optimizer import OptimizeOptions, apply_optimizations
+from .optimizer import apply_local_optimizations
+
+from pyteal.ir.tealop import TealOp
+from pyteal.ir.ops import Op
+
+from pyteal import *
+from pyteal.ast.subroutine import evaluateSubroutine
+
+import pytest
+
+
+def test_optimize_slots_basic():
+    teal = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.load, 1),
+    ]
+
+    options = OptimizeOptions(useSlotToStack=True)
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = []
+    assert result_teal == expected
+
+    options = OptimizeOptions()
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = teal
+    assert result_teal == expected
+
+
+def test_optimize_slots_iteration():
+    teal = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.store, 2),
+        TealOp(None, Op.load, 2),
+        TealOp(None, Op.load, 1),
+    ]
+
+    options = OptimizeOptions(useSlotToStack=True, iterateOptimizations=True)
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = []
+    assert result_teal == expected
+
+    options = OptimizeOptions(useSlotToStack=True)
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.load, 1),
+    ]
+    assert result_teal == expected
+
+
+def test_optimize_slots_with_dependency():
+    teal = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.store, 2),
+        TealOp(None, Op.load, 2),
+        TealOp(None, Op.load, 1),
+        TealOp(None, Op.load, 1),
+    ]
+
+    options = OptimizeOptions(useSlotToStack=True, iterateOptimizations=True)
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.load, 1),
+        TealOp(None, Op.load, 1),
+    ]
+    assert result_teal == expected
+
+    teal = [
+        TealOp(None, Op.store, 1),
+        TealOp(None, Op.store, 2),
+        TealOp(None, Op.load, 2),
+        TealOp(None, Op.load, 1),
+        TealOp(None, Op.load, 2),
+    ]
+
+    options = OptimizeOptions(useSlotToStack=True, iterateOptimizations=True)
+    result_teal = apply_local_optimizations(teal, options)
+
+    expected = teal
+    assert result_teal == expected
+
+
+def test_optimize_subroutine():
+    @Subroutine(TealType.uint64)
+    def add(a1: Expr, a2: Expr) -> Expr:
+        return a1 + a2
+
+    program = Seq(
+        [
+            If(Txn.sender() == Global.creator_address()).Then(Pop(add(Int(1), Int(2)))),
+            Approve(),
+        ]
+    )
+
+    optimize_options = OptimizeOptions()
+
+    # unoptimized
+    expected = """#pragma version 4
```

Comment:
@algoidurovic Have you considered parameterizing the tests by TEAL version?

For the PR's optimization, it seems unlikely that different TEAL versions produce different outputs.  Nevertheless, it feels to me that we ought to be able to certify optimizations for all supported versions.

Reply:
@algoidurovic That's a reasonable response.  I'm on-board to keep as is without exercising a matrix of TEAL versions.